[InstCombine] Fold shuffled intrinsic operands with constant operands

lukel97 opened the pull request:
We currently pull shuffles through binops and intrinsics, which is an important canonical form for VectorCombine to be able to scalarize vector sequences. But while binops can be folded with a constant operand, intrinsics currently require all operands to be shufflevectors.

This extends intrinsic folding to be in line with regular binops by reusing the constant "unshuffling" logic.

As far as I can tell the list of currently folded intrinsics don't require any special UB handling.

This change in combination with #138095 and #137823 fixes the following C:

```c
void max(int *x, int *y, int n) {
  for (int i = 0; i < n; i++)
    x[i] += *y > 42 ? *y : 42;
}
```

Not using the splatted vector form on RISC-V with `-O3 -march=rva23u64`:

```asm
	vmv.s.x	v8, a4
	li	a4, 42
	vmax.vx	v10, v8, a4
	vrgather.vi	v8, v10, 0
.LBB0_9:                                # %vector.body
                                        # =>This Inner Loop Header: Depth=1
	vl2re32.v	v10, (a5)
	vadd.vv	v10, v10, v8
	vs2r.v	v10, (a5)
```

i.e., it now generates

```asm
        li	a6, 42
        max	a6, a4, a6
.LBB0_9:                                # %vector.body
                                        # =>This Inner Loop Header: Depth=1
	vl2re32.v	v8, (a5)
	vadd.vx	v8, v8, a6
	vs2r.v	v8, (a5)
```

Stacked on #141287

lukel97 committed:
Precommit tests

lukel97 committed:
[InstCombine] Fold shuffled intrinsic operands with constants

dtcxzyw reviewed (approved):
The last commit LGTM.

lukel97 committed:
Fix not checking the first argument

Because we now take the mask from the first non-constant argument, it may no longer be the first overall argument.

So check all arguments now in case we have e.g.

fma(const, shuffle, shuffle)

so that the correct argument is used for the const.

nikic commented on llvm/test/Transforms/InstCombine/fma.ll:
```diff
@@ -802,6 +802,67 @@ define <2 x float> @fma_unary_shuffle_ops_narrowing(<3 x float> %x, <3 x float>
+;
+  %a = shufflevector <2 x float> %x, <2 x float> poison, <2 x i32> <i32 1, i32 0>
+  %b = shufflevector <2 x float> %y, <2 x float> poison, <2 x i32> <i32 1, i32 0>
+  %r = call <2 x float> @llvm.fma(<2 x float> <float 1.0, float 2.0>, <2 x float> %a, <2 x float> %b)
```
FWIW, I found these tests a bit confusing because the order of arguments changed. It would be more obvious to put the constants in the second arg in the first place, rather than have InstCombine canonicalize them first.

nikic commented on llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp:
```diff
@@ -1419,9 +1418,11 @@ foldShuffledIntrinsicOperands(IntrinsicInst *II,
-             m_Shuffle(m_Value(X), m_Undef(), m_Mask(Mask))))
+  auto *NonConstArg = find_if_not(II->args(), IsaPred<Constant>);
+  if (!NonConstArg ||
+      !match(NonConstArg, m_Shuffle(m_Value(X), m_Undef(), m_Mask(Mask))))
```
Might want to replace m_Undef -> m_Poison in a followup, for clarity.

nikic commented on llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp:
```diff
@@ -1430,14 +1431,21 @@ foldShuffledIntrinsicOperands(IntrinsicInst *II,
-    if (!match(II->getArgOperand(i),
-               m_Shuffle(m_Value(X), m_Undef(), m_SpecificMask(Mask))) ||
-        X->getType() != SrcTy)
+  for (unsigned i = 0, e = II->arg_size(); i != e; ++i) {
```
Could use args() here (and push_back).

nikic reviewed (approved):
LGTM

lukel97 committed:
Use args() + push_back()

lukel97 committed:
Canonicalize arg order in tests
